src/erc20.sol

```diff
@@ -97,4 +116,20 @@ contract ERC20 {
     function move(address src, address dst, uint wad) external {
         transferFrom(src, dst, wad);
     }
+
+    // --- Approve by signature ---
+    function permit(address owner, address spender, uint value, uint deadline, uint8 v, bytes32 r, bytes32 s) external {
+        require(deadline >= block.timestamp, 'cent/past-deadline');
+        bytes32 digest = keccak256(
+            abi.encodePacked(
+                '\x19\x01',
+                DOMAIN_SEPARATOR,
+                keccak256(abi.encode(PERMIT_TYPEHASH, owner, spender, value, nonces[owner]++, deadline))
+            )
+        );
+        address recoveredAddress = ecrecover(digest, v, r, s);
+        require(recoveredAddress != address(0) && recoveredAddress == owner, 'cent-erc20/invalid-sig');
```

Comment:
I am curious why a check `recoveredAddress != address(0)` is required. This is the only check i don't understand. Need to look a bit more into EIP-2612.

Reply:
Ah I see.  `address(0)` is the error return from `ecrecover`

https://solidity.readthedocs.io/en/latest/units-and-global-variables.html#mathematical-and-cryptographic-functions